### PR TITLE
Tweak `'static` suggestion code

### LINKED ISSUE #71196: Wrong suggestion of using `+ static` instead of `<'static>` for unions
Someone had this problem on discord, so creating an issue: 

```rust
pub union Test<'t> {
    i: &'t i64,
    f: &'t f64,
}
thread_local! {
    static vals: RefCell<HashMap<i32, Vec<Vec<Test>>>> = RefCell::new(HashMap::new());
}
```

Rust suggests you to: 
```
error[E0106]: missing lifetime specifier

   |
   |     static vals: RefCell<HashMap<i32, Vec<Vec<Test>>>> = RefCell::new(HashMap::new());
   |                                               ^^^^ help: consider giving it a 'static lifetime: `Test + 'static`
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be 
borrowed from
```

instead it should suggest to use `Test<'static>` and not `<Test + 'static>`

[playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d6a5bd29f891cd5f8621c6fef37c1f95)

### PR DESCRIPTION
**Pull request by estebank**
Fix #71196.

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
These suggestions look much better to me! Is this ready for review?

**Comment by estebank** [comment]
@ecstatic-morse thanks! Yes, I think I don't have anything else I can improve in the output except [for one case in particular that could be confusing, but the added suggestion helps clear it up](https://github.com/rust-lang/rust/pull/71235/files#diff-f9827d3a7a7b45558502e4e5a2bbb677R145-R155). I would also like to [clean this up](https://github.com/rust-lang/rust/pull/71235/files#diff-1bd3f65ee2d72d0225a5db83a8391165), but it can be done in a separate PR.

File: src/librustc_typeck/astconv.rs [hunk 1]
```diff
@@ -1015,18 +1015,8 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
-        let path_span = if let [segment] = &trait_ref.path.segments[..] {
-            // FIXME: `trait_ref.path.span` can point to a full path with multiple
-            // segments, even though `trait_ref.path.segments` is of length `1`. Work
-            // around that bug here, even though it should be fixed elsewhere.
-            // This would otherwise cause an invalid suggestion. For an example, look at
-            // `src/test/ui/issues/issue-28344.rs`.
-            segment.ident.span
-        } else {
-            trait_ref.path.span
-        };
```
Comment (ecstatic-morse) [inline src/librustc_typeck/astconv.rs:1027]: Out of curiosity, was this fixed elsewhere?

File: src/librustc_resolve/late/lifetimes.rs [hunk 2]
```diff
@@ -2479,29 +2454,28 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
-            true
+            false
         } else {
             db.help(&format!(
                 "this function's return type contains a borrowed value, \
                  but the signature does not say whether it is borrowed from {}",
                 m
             ));
-            true
+            false
```
Comment (ecstatic-morse) [inline src/librustc_resolve/late/lifetimes.rs:2478]: It seems that the semantics of the return value of `report_elision_failure` have changed. What does the `bool` indicate here? Would you mind documenting it?

**Review by ecstatic-morse — APPROVED** [review]
Sweet! As I mentioned above, these errors read much better to me. I like how the fix is written out in full, especially in the more complex cases. The logic seems cleaner than the old version as well. I left a few requests for clarification but overall I think this is good.

**Comment by ecstatic-morse** [comment]
r=me with comment about meaning of return value for `report_elision_failure`.

File: src/librustc_typeck/astconv.rs (hunk 1 above)
Comment (estebank) [reply src/librustc_typeck/astconv.rs:1027]: Partly. I think this used to affect multiple tests but the only case that appears now is at https://github.com/rust-lang/rust/pull/71235/files#diff-1bd3f65ee2d72d0225a5db83a8391165.

**Comment by JohnTitor** [comment]
Seems failed in rollup: https://github.com/rust-lang/rust/pull/71377#issuecomment-617484400
@bors r-

**Pull request merged**